### ISSUE TITLE
Broken bootstrapping due to missing include in 1.1.3

### ISSUE THREAD

**Issue by mhthies** [just-buildsystem/justbuild#36]
Bootrapping justbuild 1.1.3 for packaging on Arch Linux fails with the following error during the first stage:

```text
Running ['/usr/bin/clang++', '-target', 'x86_64-linux-gnu', '-O2', '-DNDEBUG', '-std=c++20', '-Wall', '-Wextra', '-Wpedantic', '-Wsign-conversion', '-Werror', '-pedantic-errors', '-Wno-error', '@protobuf.cflags', '@grpc++.cflags', '@nlohmann_json.cflags', '@fmt.cflags', '@gsl.cflags', '-I', 'work', '-isystem', 'include', '-c', 'work/src/buildtool/build_engine/expression/expression.cpp', '-o', 'work/src/buildtool/build_engine/expression/expression.o'] with env {'PATH': '/bin:/sbin:/usr/bin:/usr/sbin', 'PKG_CONFIG_PATH': '/home/...:/usr/lib/pkgconfig:/usr/share/pkgconfig'} for action '27ef66d6dbe78cb9905a728a05af3baee700acaf'
In file included from work/src/buildtool/execution_api/execution_service/operation_cache.cpp:15:
work/src/buildtool/execution_api/execution_service/operation_cache.hpp:66:14: error: no type named 'unique_lock' in namespace 'std'
        std::unique_lock lock{mutex_};
        ~~~~~^
work/src/buildtool/execution_api/execution_service/operation_cache.cpp:38:14: error: no type named 'unique_lock' in namespace 'std'
        std::unique_lock ulock{mutex_};
        ~~~~~^
```

Built with clang 15.0.7.

The error seems to be caused by a missing include of `<mutex>` in src/buildtool/execution_api/execution_service/operation_cache.hpp. It can be fixed with the following patch:

```diff
diff --git a/src/buildtool/execution_api/execution_service/operation_cache.hpp b/src/buildtool/execution_api/execution_service/operation_cache.hpp
index 97b137f..25ca49a 100644
--- a/src/buildtool/execution_api/execution_service/operation_cache.hpp
+++ b/src/buildtool/execution_api/execution_service/operation_cache.hpp
@@ -16,6 +16,7 @@
 #define OPERATION_CACHE_HPP
 
 #include <atomic>
+#include <mutex>
 #include <optional>
 #include <shared_mutex>
 #include <string>
```

**Comment by mhthies** [comment]
Now, I'm a bit confused about this error: I can reproduce it with the justbuild 1.1.2 codebase as well, but I didn't encounter it, when I tested my Archl Linux package recipe for version 1.1.2 last time. And there hasn't been a update to libstdc++ in the meantime, as far as I can tell.

**Comment by aehlig** [comment]
It is indeed an interesting question why
- this hasn't been observed, and
- the problem did not occur on the architectures we tested it (admittedly, not including Arch Linux).

Nevertheless, it definitely is an error to use `std::unique_lock` without including `mutex`;  the standard does not guarantee that `shared_mutex` also pulls in `mutex`. So, if you agree, I will include that patch in your name to the code base. Should we do a 1.1.4 release for that fix?

**Comment by mhthies** [comment]
I also really don't know what I have changed to my setup to expose this error.

Of course, feel free to include the patch with my name. A 1.1.4 bug fix release might be useful to allow building an Arch Linux package with the updates from version 1.1.3. However, I'm currently facing another issue with the 1.1.3 package build, while linking the native just-mr executable. I'll try to analyze this first and report it here, so it might be fixed in the same release – if a patch to the codebase is required for this issue.

**Comment by mhthies** [comment]
Ah, I think it's the update of abseil-cpp from version 20230125.1 to 20230125.3 that makes this missing include show up.

**Comment by aehlig** [comment]
c87d2ac032f682f489e23d5950cd3166f3e303a2 committed on master and will be cherry-picked to [stable-1.1](https://github.com/just-buildsystem/justbuild/tree/stable-1.1).

**Comment by aehlig** [comment]
Fix included in release [1.1.4](https://github.com/just-buildsystem/justbuild/releases/tag/v1.1.4).
